### PR TITLE
Add support for local suppression inside formatting

### LINKED ISSUE #5547: `EnumEntryNameCase` cannot be suppressed on (enum) class or declaration level
## Expected Behavior

I should be able to suppress `EnumEntryNameCase` via

```
@Suppress("EnumEntryNameCase")
enum class SpdxLicense(
    val id: String,
    val fullName: String,
    val deprecated: Boolean = false
) {
    _0BSD("0BSD", "BSD Zero Clause License");
}
```

or

```
enum class SpdxLicense(
    val id: String,
    val fullName: String,
    val deprecated: Boolean = false
) {
    @Suppress("EnumEntryNameCase")
    _0BSD("0BSD", "BSD Zero Clause License");
}
```

## Observed Behavior

Suppressing this way has no effect, the inspection hint is already being reporter. However, suppressing on the *file*-level works:

```
@file:Suppress("EnumEntryNameCase")

enum class SpdxLicense(
    val id: String,
    val fullName: String,
    val deprecated: Boolean = false
) {
    _0BSD("0BSD", "BSD Zero Clause License");
}
```

## Steps to Reproduce

See code examples above.

## Context

I ran into this while upgrading to Detekt 1.22 in https://github.com/oss-review-toolkit/ort/pull/6096.

## Your Environment

* Version of detekt used: 1.22
* Version of Gradle used (if applicable): 7.5.1
* Gradle scan link (add `--scan` option when running the gradle task): n/a
* Operating System and version: Windows 11
* Link to your project (if it's a public repository): https://github.com/oss-review-toolkit/ort

### LINKED ISSUE #5566: Cannot suppress `TrailingCommaOnDeclarationSite ` on enum with trailing semicolon
## Expected Behavior
When using `TrailingCommaOnDeclarationSite`, if an enum has a semicolon to separate the last item, KtLint adds a comma before. I reported the error in pinterest/ktlint/issues/1714 and it seems that the KtLint suppress is working, but the suppress for detekt is not.

```kotlin
enum class Example {
    A,
    B;

    fun something() = "blah"
}

enum class Example2 {
    A,
    B;

    companion object {
        val values = listOf(A, B)
    }
}
```

is formatted into

```kotlin
enum class Example {
    A,
    B,
    ;

    fun something() = "blah"
}

enum class Example2 {
    A,
    B,
    ;

    companion object {
        val values = listOf(A, B)
    }
}
```

## Observed Behavior
If I add `@Suppress("TrailingCommaOnDeclarationSite")` before the declaration of the enum, the error won't be suppressed and is still formatting the enum.

## Steps to Reproduce
Using this configuration of detekt

```yaml
TrailingCommaOnCallSite:
  active: true
  autoCorrect: true
  useTrailingCommaOnCallSite: true
TrailingCommaOnDeclarationSite:
  active: true
  autoCorrect: true
  useTrailingCommaOnDeclarationSite: true
```

and

```kotlin
@Suppress("TrailingCommaOnDeclarationSite")
enum class Example {
    A,
    B;

    fun something() = "blah"
}
```

## Context
This happens with detekt 1.22.0. The error should be suppressed, otherwise it will format the enum adding a trailing comma before the semicolon.

### PR DESCRIPTION
**Pull request by cortinico**
Fixes #5547
Fixes #5566
Related to #1999 also

This introduced the possibility to do more fine grained suppression on `detekt-formatting` rules. Specifically I'm dropping the custom logic we apply to the `Entity` we receive and just pass over the `PsiElement` that KtLint passes us.

Some of the location in the tests changed due to this, so this is a disruptive change at least for baselines (as the signature also changes). If we decide to merge it we should flag it in the release notes.

At the same time with KtLint you could only suppress at the `@file:` level so, aside from baseline, this won't disrupt other workflows.

### CODE REVIEW

File: detekt-formatting/src/test/kotlin/io/gitlab/arturbosch/detekt/formatting/MaximumLineLengthSpec.kt [hunk 1]
```diff
@@ -55,14 +56,13 @@ class MaximumLineLengthSpec {
...
 
         // Note that KtLint's MaximumLineLength rule, in contrast to detekt's MaxLineLength rule, does not report
         // exceeded lines in block comments.
-        assertThat(findings).hasSize(2)
-
-        assertThat(findings[0].entity.location.source.line).isEqualTo(8)
-        assertThat(findings[1].entity.location.source.line).isEqualTo(14)
+        io.gitlab.arturbosch.detekt.test.assertThat(findings).hasSize(2)
+        io.gitlab.arturbosch.detekt.test.assertThat(findings[0]).hasSourceLocation(7, 8)
+        io.gitlab.arturbosch.detekt.test.assertThat(findings[1]).hasSourceLocation(13, 12)
```
Comment (cortinico) [inline detekt-formatting/src/test/kotlin/io/gitlab/arturbosch/detekt/formatting/MaximumLineLengthSpec.kt:65]: I have a follow-up PR to clean all the `detekt-formatting` from using non detekt's `assertThat` function here: https://github.com/cortinico/detekt/commit/d9a4ea9a90bf9da6b100aa8a0848cbb8cb3b07c8

**Comment by cortinico** [comment]
Can someone take a look at this one?

**Review by BraisGabin — APPROVED** [review]

File: detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRule.kt [hunk 2]
```diff
@@ -41,32 +36,6 @@ abstract class FormattingRule(config: Config) : Rule(config) {
     val runAsLateAsPossible
         get() = RunAsLateAsPossible in wrapping.visitorModifiers
 
-    private val emit = { offset: Int, message: String, canBeAutoCorrected: Boolean ->
-        val (line, column) = positionByOffset(offset)
-        val location = Location(
```
Comment (chao2zhang) [inline detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRule.kt:46]: Would you mind adding a unit test for the location if such test has not existed

**Review by chao2zhang — APPROVED** [review]

File: detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRule.kt (hunk 2 above)
Comment (cortinico) [reply detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRule.kt:46]: I've updated the tests to follow the logic here (i.e. previously we had a test that was checking file suppression as valid and local suprresion as not. Now the test checks that both suppressions are valid).

**Pull request merged**